File: code/stable/dblpendulum/src/python/Calculations.py
## \file Calculations.py
# \author Dong Chen
# \brief Provides functions for calculating the outputs
import math

import scipy.integrate

## \brief Calculates dependent variables (rad)
# \param m_1 the mass of the first object (kg)
# \param m_2 the mass of the second object (kg)
# \param L_2 the length of the second rod (m)
# \param L_1 the length of the first rod (m)
# \return dependent variables (rad)
def func_theta(m_1, m_2, L_2, L_1):
    def f(t, theta):
        return [theta[1], (-9.8 * (2.0 * m_1 + m_2) * math.sin(theta[0]) - m_2 * 9.8 * math.sin(theta[0] - 2.0 * theta[2]) - 2.0 * math.sin(theta[0] - theta[2]) * m_2 * (theta[3] ** 2.0 * L_2 + theta[1] ** 2.0 * L_1 * math.cos(theta[0] - theta[2]))) / (L_1 * (2.0 * m_1 + m_2 - m_2 * math.cos(2.0 * theta[0] - 2.0 * theta[2]))), theta[3], 2.0 * math.sin(theta[0] - theta[2]) * (theta[1] ** 2.0 * L_1 * (m_1 + m_2) + 9.8 * (m_1 + m_2) * math.cos(theta[0]) + theta[3] ** 2.0 * L_2 * m_2 * math.cos(theta[0] - theta[2])) / (L_2 * (2.0 * m_1 + m_2 - m_2 * math.cos(2.0 * theta[0] - 2.0 * theta[2])))]
    
    r = scipy.integrate.ode(f)
    r.set_integrator("dopri5", atol=1.0e-6, rtol=1.0e-6)
    r.set_initial_value([1.3463968515384828, 0.0, 2.356194490192345, 0.0], 0.0)
    theta = [[1.3463968515384828, 0.0, 2.356194490192345, 0.0][0]]
    while r.successful() and r.t < 20.0:
        r.integrate(r.t + 1.0e-3)
        theta.append(r.y[0])
    
    return theta

File: code/stable/dblpendulum/src/java/DblPendulum/Calculations.java
package DblPendulum;

/** \file Calculations.java
    \author Dong Chen
    \brief Provides functions for calculating the outputs
*/
import java.util.ArrayList;
import org.apache.commons.math3.ode.FirstOrderIntegrator;
import org.apache.commons.math3.ode.nonstiff.DormandPrince54Integrator;

public class Calculations {
    
    /** \brief Calculates dependent variables (rad)
        \param m_1 the mass of the first object (kg)
        \param m_2 the mass of the second object (kg)
        \param L_2 the length of the second rod (m)
        \param L_1 the length of the first rod (m)
        \return dependent variables (rad)
    */
    public static ArrayList<Double> func_theta(double m_1, double m_2, double L_2, double L_1) {
        ArrayList<Double> theta;
        ODEStepHandler stepHandler = new ODEStepHandler();
        ODE ode = new ODE(m_1, m_2, L_1, L_2);
        double[] curr_vals = {1.3463968515384828, 0.0, 2.356194490192345, 0.0};
        
        FirstOrderIntegrator it = new DormandPrince54Integrator(1.0e-3, 1.0e-3, 1.0e-6, 1.0e-6);
        it.addStepHandler(stepHandler);
        it.integrate(ode, 0.0, curr_vals, 20.0, curr_vals);
        theta = stepHandler.theta;
        
        return theta;
    }
}
